classes/Services/PhpVersionResolverService.php

```diff
@@ -0,0 +1,254 @@
+            throw new UpgradeException('Unable to retrieve latest 1.7 release of Prestashop.');
+        }
+
+        if ($currentPhpVersion < 70200) {
+            $release = $releasesFromChannelFile['1.7'];
+
+            // current version is superior or equal
+            if (version_compare($this->currentPsVersion, $release->getVersion(), '>=')) {
+                return null;
+            }
+
+            return $releasesFromChannelFile['1.7'];
+        }
+
+        $validReleases = [];
+
+        foreach ($this->distributionApiService->getReleases() as $release) {
+            if ($release->getStability() !== 'stable') {
+                continue;
+            }
+
+            // current version is superior or equal
+            if (version_compare($this->currentPsVersion, $release->getVersion(), '>=')) {
+                continue;
+            }
+
+            $versionMin = VersionUtils::getPhpVersionId($release->getPhpMinVersion());
+            $versionMax = VersionUtils::getPhpVersionId($release->getPhpMaxVersion());
+
+            $versionMinWithoutPatch = VersionUtils::getPhpMajorMinorVersionId($versionMin);
+            $versionMaxWithoutPatch = VersionUtils::getPhpMajorMinorVersionId($versionMax);
+
+            // verify php compatibility
+            if ($currentPhpVersion >= $versionMinWithoutPatch && $currentPhpVersion <= $versionMaxWithoutPatch) {
+                // verify channel.xml matching
+                $branch = VersionUtils::splitPrestaShopVersion($release->getVersion())['minor'];
+                if (isset($releasesFromChannelFile[$branch])) {
+                    $releaseNote = $releasesFromChannelFile[$branch]->getReleaseNoteUrl();
+                    $release->setReleaseNoteUrl($releaseNote);
+                    $validReleases[] = $release;
+                }
+            }
+        }
+
+        /** @var PrestashopRelease $release */
+        $release = array_reduce($validReleases, function ($carry, $item) {
+            if ($carry === null || version_compare($item->getVersion(), $carry->getVersion()) > 0) {
+                return $item;
+            }
+
+            return $carry;
+        });
+
+        return $release;
+    }
+
+    /**
+     * @return PrestashopRelease[]
+     */
+    private function getPrestashop17Requirements(): array
+    {
+        return [
+            new PrestashopRelease('1.7.0.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.1.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.2.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.3.0', 'stable', '7.1', '5.4'),
+            new PrestashopRelease('1.7.4.0', 'stable', '7.1', '5.6'),
+            new PrestashopRelease('1.7.5.0', 'stable', '7.2', '5.6'),
+            new PrestashopRelease('1.7.6.0', 'stable', '7.2', '5.6'),
+            new PrestashopRelease('1.7.7.0', 'stable', '7.3', '5.6'),
+            new PrestashopRelease('1.7.8.0', 'stable', '7.4', '5.6'),
```

Comment:
```suggestion
            '1.7.0' => new PrestashopRelease('1.7.0.0', 'stable', '7.1', '5.4'),
            '1.7.1' => new PrestashopRelease('1.7.1.0', 'stable', '7.1', '5.4'),
            // [...]
            '1.7.8' => new PrestashopRelease('1.7.8.0', 'stable', '7.4', '5.6'),
```

This would remove the foreach you have in the caller.

```php
            $targetMinorVersion = VersionUtils::splitPrestaShopVersion($targetVersion)['minor'];

            $requirements = $this->getPrestashop17Requirements();
            if (isset($requirements[$targetMinorVersion])) {
                $range = [
                    'php_min_version' => $requirements[$targetMinorVersion]->getPhpMinVersion(),
                    'php_max_version' => $requirements[$targetMinorVersion]->getPhpMaxVersion(),
                ];
            }
```

I prefer the previous version, the array allowed you to assign the compatibility range directly.